### PR TITLE
refactor: bug fixes and refactor for vlm

### PR DESCRIPTION
**Pull request by mickqian**
## Motivation

1. a general vlm embed routine
2. sglang image processors architecture improvement
3. fix the bug of registering customized image processor to transformers
4. add some consistency to vlm api

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/managers/mm_utils.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        )
+
+        inputs_embeds = inputs_embeds.masked_scatter(
+            special_image_mask,
+            image_embedding.to(inputs_embeds.device, inputs_embeds.dtype),
+        )
+    return inputs_embeds
+
+
+def embed_image_embedding(
```
Comment (mickqian) [inline python/sglang/srt/managers/mm_utils.py:250]: After this modification, the result of mmmu benchmark of sglang goes from:
|     |  accuracy  |  time elapsed (s) |
| --- | --- | --- |
|   Qwen2.5-VL-7B-Instruct |    0.462  | 229.1  |
|   gemma-3-4b-it | 0.409   | 226.4   |

to:


|     |  accuracy  |  time elapsed (s)  |
| --- | --- | --- |
|   Qwen2.5-VL-7B-Instruct |    0.477   | 230.2  |
|   gemma-3-4b-it |  0.409    | 202.3  |

**Comment by mickqian** [comment]
This should fix #4607 #4159 and many other similar issues I can't count

**Comment by zhaochenyang20** [comment]
@mickqian I've rerun the CI. thanks! minicpm-o is of high-priority qvq

**Comment by zhaochenyang20** [comment]
@mickqian this is not ready to review right? mick, if ready, please ping yi and me

**Comment by mickqian** [comment]
@yizhang2077 @zhaochenyang20 ready now. vlm tests all passed. Thanks

File: test/srt/run_suite.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ class TestFile:
         TestFile("test_vertex_endpoint.py", 31),
         TestFile("test_vision_chunked_prefill.py", 223),
         TestFile("test_vlm_accuracy.py", 60),
-        TestFile("test_vision_openai_server.py", 344),
+        TestFile("test_vision_openai_server.py", 546),
```
Comment (yizhang2077) [inline test/srt/run_suite.py]: why do we need increase test time here?

File: python/sglang/srt/configs/utils.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    PretrainedConfig,
+    ProcessorMixin,
+)
+
+
+def register_image_processor(
+    config: Type[PretrainedConfig], image_processor: Type[BaseImageProcessor]
+):
+    """
+    register customized hf image processor while removing hf impl
```
Comment (yizhang2077) [inline python/sglang/srt/configs/utils.py:16]: I don't understand why do we need add register function here

File: python/sglang/srt/models/gemma3_mm.py [hunk 4]
```diff
@@ -258,10 +259,10 @@ def prepare_attn_masks(
         kwargs["local_attn_masks"] = local_attn_masks
         return kwargs
 
-    def get_input_embeddings(self):
+    def get_input_embeddings(self) -> nn.Embedding:
         return self.language_model.get_input_embeddings()
 
-    def get_image_features(self, pixel_values: torch.Tensor):
+    def get_image_feature(self, image_input: ImageInputs):
```
Comment (yizhang2077) [inline python/sglang/srt/models/gemma3_mm.py:271]: why do we change from `get_image_features` -> `get_image_feature`

File: python/sglang/srt/models/minicpmv.py [hunk 5]
```diff
@@ -836,46 +801,6 @@ def _parse_and_validate_inputs(
...
-        pixel_values_flat: List[torch.Tensor] = []
-        tgt_sizes_flat: List[torch.Tensor] = []
-        for pixel_b, tgt_b in zip(pixel_values, tgt_sizes):
-            if len(pixel_b) != len(tgt_b):
-                raise ValueError(
-                    "Inconsistent N lengths, found: " f"{len(pixel_b)} vs {len(tgt_b)}"
-                )
-
-            for pixel_n, tgt_n in zip(pixel_b, tgt_b):
-                pixel_values_flat += pixel_n
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmv.py:864]: why do we remove lots of content here?

File: python/sglang/srt/models/qwen2_5_vl.py [hunk 6]
```diff
@@ -714,4 +645,4 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
 
 
 EntryClass = [Qwen2_5_VLForConditionalGeneration]
-AutoModel.register(Qwen2_5_VLConfig, Qwen2_5_VLForConditionalGeneration)
+AutoModel.register(Qwen2_5_VLConfig, Qwen2_5_VLForConditionalGeneration, exist_ok=True)
```
Comment (yizhang2077) [inline python/sglang/srt/models/qwen2_5_vl.py]: why do we need add exist_ok here?

**Review by yizhang2077 — COMMENTED** [review]
I add some basic comments here, continue to reviewing

File: test/srt/run_suite.py (hunk 2 above)
Comment (mickqian) [reply test/srt/run_suite.py]: the original one is obsolete, I reran it

File: python/sglang/srt/configs/utils.py (hunk 3 above)
Comment (mickqian) [reply python/sglang/srt/configs/utils.py:16]: Because previous register function introduces `is already used by transformers` error. This can be useful, if we expect a lot of backport of  AutoProcessors

File: python/sglang/srt/models/gemma3_mm.py (hunk 4 above)
Comment (mickqian) [reply python/sglang/srt/models/gemma3_mm.py:271]: Just to get a universal api across all vlms: all renamed to `get_image_feature`

File: python/sglang/srt/models/minicpmv.py (hunk 5 above)
Comment (mickqian) [reply python/sglang/srt/models/minicpmv.py:864]: I move it into the Processor, the logic seems tedious here in model file

File: python/sglang/srt/models/qwen2_5_vl.py (hunk 6 above)
Comment (mickqian) [reply python/sglang/srt/models/qwen2_5_vl.py]: replied [here](https://github.com/sgl-project/sglang/pull/4661#discussion_r2008993437)

File: python/sglang/srt/conversation.py [hunk 7]
```diff
@@ -653,7 +653,7 @@ def generate_chat_conv(
     Conversation(
         name="gemma-it",
         system_message="You are a helpful assistant.",
-        system_template="<bos><start_of_turn>user{system_message}\n\n",
+        system_template="<start_of_turn>user{system_message}\n\n",
```
Comment (mickqian) [inline python/sglang/srt/conversation.py:656]: Since the transformers' update, the tokenizer from hf add bos as default, bos is removed here

File: test/srt/test_vlm_accuracy.py [hunk 8]
```diff
@@ -183,22 +189,21 @@ async def test_encode_output(self):
             )
             hf_output = hf_output.squeeze(0)
 
-        with torch.no_grad():
+            # sglang
             model = self.get_sglang_model()
             input_ids = inputs["input_ids"].to(self.device).flatten()
-            image_inputs = model._parse_and_validate_inputs(
+            print(inputs)
```
Comment (yizhang2077) [inline test/srt/test_vlm_accuracy.py]: remove print here

File: python/sglang/srt/managers/schedule_batch.py [hunk 9]
```diff
@@ -204,50 +203,94 @@ def from_dict(obj: dict):
...
                 setattr(ret, arg, obj[arg])
 
+        # validate
+        assert (
+            isinstance(ret.pixel_values, torch.Tensor)
+            or isinstance(ret.pixel_values, np.ndarray)
+            or isinstance(ret.pixel_values, list)
+        )
+
+        # if ret.images_emb_mask is None:
```
Comment (yizhang2077) [inline python/sglang/srt/managers/schedule_batch.py]: remove comment here

**Comment by yizhang2077** [comment]
LGTM, great code refactoring! After CI pass, this PR can be merged cc @zhaochenyang20

**Review by yizhang2077 — APPROVED** [review]

**Comment by zhaochenyang20** [comment]
@mickqian @yizhang2077 great!
